std/experimental/allocator/building_blocks/bitmapped_block.d

```diff
@@ -229,571 +173,1462 @@ struct BitmappedBlock(size_t theBlockSize, uint theAlignment = platformAlignment
+                            cast(uint) (64 - msbIdx - blocks),
+                            63 - msbIdx);
+                    }
+                    return true;
+                }
+                else
+                {
+                    static if (isShared)
+                    {
+                        ulong localControl = _control.rep[wordIdx];
+                        localControl &= ulong.max << (64 - msbIdx);
+                        _control.rep[wordIdx] = localControl;
+                    }
+                    else
+                    {
+                        _control.rep[wordIdx] &= ulong.max << (64 - msbIdx);
+                    }
+                    blocks -= 64 - msbIdx;
+                    ++wordIdx;
+                    msbIdx = 0;
+                }
+            }
+
+            // Stage 2: reset one word at a time
+            for (; blocks >= 64; blocks -= 64)
+            {
+                _control.rep[wordIdx++] = 0;
+            }
+
+            // Stage 3: deal with leftover bits, if any
+            assert(wordIdx <= _control.rep.length);
+            if (blocks)
+            {
+                static if (isShared)
+                {
+                    ulong localControl = _control.rep[wordIdx];
+                    localControl &= ulong.max >> blocks;
+                    _control.rep[wordIdx] = localControl;
+                }
+                else
+                {
+                    _control.rep[wordIdx] &= ulong.max >> blocks;
+                }
+            }
+            return true;
+        }
+
+        // Since the lock is not pure, only the single threaded version is pure
+        static if (isShared)
+        {
+            nothrow @nogc
+            bool deallocateAll()
+            {
+                lock.lock();
+                scope(exit) lock.unlock();
+
+                (cast(BitVector) _control)[] = 0;
+                _startIdx = 0;
+                return true;
+            }
+        }
+        else
+        {
+            pure nothrow @nogc
+            bool deallocateAll()
+            {
+                _control[] = 0;
+                _startIdx = 0;
+                return true;
+            }
+        }
+
+        // Since the lock is not pure, only the single threaded version is pure
+        static if (isShared)
+        {
+            nothrow @safe @nogc
+            Ternary empty()
+            {
+                lock.lock();
+                scope(exit) lock.unlock();
+
+                return emptyImpl();
+            }
+        }
+        else
+        {
+            pure nothrow @safe @nogc
+            Ternary empty()
+            {
+                return Ternary(_control.allAre0());
+            }
+        }
+
+        pure nothrow @trusted @nogc
+        private Ternary emptyImpl()
+        {
+            return Ternary((cast(BitVector) _control).allAre0());
+        }
+
+        // Debug helper
+        private void dump()
```

Comment:
`debug(StdBitmapped)` or `version(StdBitmappedDebug)` to avoid that the compiler parses every time the module is used.